lib/internal/observable.js

```diff
@@ -0,0 +1,898 @@
+const {
+  ReflectConstruct,
+  Symbol,
+  SymbolAsyncIterator,
+  SymbolIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+
+const {
+  AbortController,
+  AbortSignal,
+} = require('internal/abort_controller');
+
+const {
+  validateFunction,
+  validateObject,
+  validateAbortSignal,
+} = require('internal/validators');
+
+const {
+  codes: {
+    ERR_ILLEGAL_INVOCATION,
+    ERR_INVALID_ARG_TYPE,
+    ERR_METHOD_NOT_IMPLEMENTED,
+  },
+} = require('internal/errors');
+
+const {
+  createDeferredPromise,
+} = require('internal/util');
+
+const kClose = Symbol('kClose');
+const kSignal = Symbol('kSignal');
+const kState = Symbol('kState');
+const kObserverMarker = Symbol('kObserverMarker');
+const kTeardown = Symbol('kTeardown');
+
+/**
+ * @typedef {Object} SubscribeOptions
+ * @property {AbortSignal} [signal]
+ */
+
+/**
+ * @callback SubscriberNextCallback
+ * @param {any} value
+ * @returns {void}
+ */
+
+/**
+ * @callback SubscriberErrorCallback
+ * @param {Error} error
+ * @returns {void}
+ */
+
+/**
+ * @callback SubscriberCompleteCallback
+ * @returns {void}
+ */
+
+/**
+ * @callback SubscriberCallback
+ * @param {Subscriber} subscriber
+ */
+
+/**
+ * @callback SubscriptionObserverCallback
+ * @param {any} value
+ * @returns {void}
+ */
+
+/**
+ * @callback VoidFunction
+ * @returns {void}
+ */
+
+/**
+ * @typedef {Object} SubscriptionObserver
+ * @property {SubscriptionObserverCallback} [next]
+ * @property {SubscriptionObserverCallback} [error]
+ * @property {VoidFunction} [complete]
+ */
+
+/**
+ * @callback Mapper
+ * @param {any} element
+ * @param {number} index
+ * @returns {any}
+ */
+
+/**
+ * @callback Predicate
+ * @param {any} element
+ * @param {number} index
+ * @returns {boolean}
+ */
+
+/**
+ * @callback Reducer
+ * @param {any} accumulator
+ * @param {any} element
+ * @returns {any}
+ */
+
+class Subscriber {
+  constructor() {
```

Comment:
The "better" pattern in our code is to pass a symbol brand here like `kInternal` instead of `ReflectConstruct` which doesn't have the performance penalty for the happy case